Userland/Services/PreviewServer/Cache.h

```diff
@@ -0,0 +1,140 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/HashMap.h>
+#include <AK/LexicalPath.h>
+#include <AK/Singleton.h>
+#include <LibCore/EventLoop.h>
+#include <LibCore/SharedCircularQueue.h>
+#include <LibCore/StandardPaths.h>
+#include <LibCrypto/Hash/SHA2.h>
+#include <LibGfx/ShareableBitmap.h>
+#include <LibThreading/ConditionVariable.h>
+#include <LibThreading/Thread.h>
+
+namespace PreviewServer {
+
+using Hash = Crypto::Hash::SHA512::DigestType;
+
+}
+
+namespace AK {
+template<>
+struct Traits<PreviewServer::Hash> : public AK::GenericTraits<PreviewServer::Hash> {
+    using Hash = PreviewServer::Hash;
+    static constexpr unsigned hash(Hash value)
+    {
+        static_assert(popcount(Hash::Size) == 1);
+        Array<u32, Hash::Size> partial_hashes;
+        for (size_t i = 0; i < partial_hashes.size(); ++i)
+            partial_hashes[i] = Traits<u8>::hash(value.bytes()[i]);
+
+        size_t hash_count = partial_hashes.size();
+        // Iteratively combine pairs of hashes.
+        while (hash_count > 1) {
+            for (size_t i = 0; i < hash_count; i += 2) {
+                auto left = partial_hashes[i];
+                auto right = partial_hashes[i + 1];
+                partial_hashes[i / 2] = pair_int_hash(left, right);
+            }
+            hash_count /= 2;
+        }
+        return partial_hashes[0];
+    }
+};
+}
+
+namespace PreviewServer {
+
+// TODO: Make the shareable bitmaps volatile while we're not sending them out,
+//       so that cache entries can be purged by the Kernel under memory pressure.
+// TODO: Keep disk cache and in-memory cache small by regularly removing old entries.
+struct ValidCacheEntry {
+    ValidCacheEntry()
+        : path(DeprecatedString::empty())
+    {
+    }
+    ValidCacheEntry(Gfx::ShareableBitmap preview, LexicalPath path)
+        : preview(move(preview))
+        , path(move(path))
+    {
+    }
+    Gfx::ShareableBitmap preview;
+    LexicalPath path;
+};
+
+// This struct MUST be incompatible with ErrorOr.
+// Otherwise, Promise will "flatten" our contained error and crash because it expected a successful result.
+struct CacheEntry {
+    CacheEntry()
+        : entry(ValidCacheEntry {})
+    {
+    }
+    explicit CacheEntry(ErrorOr<ValidCacheEntry>&& entry)
+        : entry(move(entry))
+    {
+    }
+    explicit CacheEntry(CacheEntry const& cache_entry)
+        : entry(cache_entry.entry.copy())
+    {
+    }
+    CacheEntry(CacheEntry&&) = default;
+    CacheEntry& operator=(CacheEntry&&) = default;
+    CacheEntry(Error&& error)
+        : entry(move(error))
+    {
+    }
+    CacheEntry(ValidCacheEntry&& valid_entry)
+        : entry(move(valid_entry))
+    {
+    }
+    ErrorOr<ValidCacheEntry> entry;
+};
+
+using CachePromise = Core::Promise<CacheEntry>;
+
+static LexicalPath s_cache_directory = LexicalPath::join(Core::StandardPaths::home_directory(), ".cache/preview"sv);
```

Comment:
probably just a note for the future: not sure whether contents of this directory should be more or less global and owned by the server instead. (avoid work in multiple users scenario—knowing that the previews were generated earlier shouldnʼt be a security threat but… canʼt guarantee that! 😅)

Reply:
my idea here was it is the system that generates those and has full control over it—not any of the OS users. this way generated thumbnails can be shared and generated once for all the users.

still not sure how to make proper security guarantees. (so thatʼs just more of a loose idea than anything else)